File: 27-01-23/serialize_and_deserialize.py
class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "N"
        res=[]
        l=[root]
        res=[str(root.val)]
        while(l):
            length=len(l)
            for i in range(0,length):
                cur=l[0]
                l.remove(cur)
                if not cur:
                    res.append("N")
                res.append(str(cur.val))
                l.append(cur.left)
                l.append(cur.right)
        return ",".join(res)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        val=iter(data.split(","))
        def des():
            if val=="N":
                return None
            node=TreeNode(str(val))
            node.left=des()
            node.right=des()
            return node
        return des()

File: Feb_3/serialize_and_deserialize.java
public class Codec {
    public String serialize(TreeNode root) {
        if(root==null){
            return "N";
        }
        return String.valueOf(root.val)+","+serialize(root.left)+","+serialize(root.right);
    }
    int ind=0;
    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] arr=data.split(",");
        return pre(arr);
    }
    TreeNode pre(String[] arr){
        if(arr[ind].equals("N")){
            ind++;
            return null;
        }
        TreeNode node=new TreeNode(Integer.parseInt(arr[ind++]));
        node.left=pre(arr);
        node.right=pre(arr);
        return node;
    }

    // Encodes a tree to a single string.
    
}
